python/ray/util/sgd/v2/tests/test_trainer.py

```diff
@@ -206,6 +206,62 @@ def train_mismatch():
             trainer.run(train)
 
 
+def test_checkpoint(ray_start_2_cpus):
+    config = TestConfig()
+
+    def train_func():
+        assert sgd.load_checkpoint() is None
+        for i in range(3):
+            sgd.save_checkpoint(epoch=i)
+        return 1
+
+    trainer = Trainer(config, num_workers=2)
+    trainer.start()
+    trainer.run(train_func)
+    checkpoint = trainer.get_latest_checkpoint()
+
+    assert checkpoint is not None
+    assert checkpoint["epoch"] == 2
+
+    def train_func_checkpoint():
+        checkpoint = sgd.load_checkpoint()
+        assert checkpoint is not None
+        assert checkpoint["epoch"] == 2
+
+        for i in range(checkpoint["epoch"], 5):
+            sgd.save_checkpoint(epoch=i)
+        return 1
+
+    trainer.run(train_func_checkpoint, checkpoint=checkpoint)
+    checkpoint = trainer.get_latest_checkpoint()
+
+    assert checkpoint is not None
+    assert checkpoint["epoch"] == 4
+
+
+def test_load_checkpoint(ray_start_2_cpus):
+    config = TestConfig()
+
+    def train_func_checkpoint():
+        checkpoint = sgd.load_checkpoint()
+        assert checkpoint is not None
+        assert checkpoint["epoch"] == 3
+
+        result = []
+        for i in range(checkpoint["epoch"], 5):
+            result.append(i)
+        return result
+
+    trainer = Trainer(config, num_workers=2)
+    trainer.start()
+    result = trainer.run(train_func_checkpoint, checkpoint={"epoch": 3})
+
+    assert result is not None
+    assert len(result) == 2
+    assert result[0] == [3, 4]
+    assert result[1] == [3, 4]
+
+
```

Comment:
Can you add more tests here, particularly for the failure cases?

- Mismatch in number of checkpoint calls/report calls
- Adding checkpointing to the `test_fast_slow`
- What happens if there is worker failure? Particularly if one of the worker fails in between a `save_checkpoint` and a `report` call. Does the latest checkpoint get saved properly? Also what about if there is a failure in between a `save_checkpoint` and the training function finishing.

Reply:
Added!